### ISSUE TITLE
plugin-cloud-storage: Can't resolve 'fs' error with cloud-storage-plugin s3 adapter

### ISSUE THREAD

**Issue by haliacmon** [payloadcms/payload#4215]
### Link to reproduction

https://github.com/haliacmon/payload-reproduction

### Describe the Bug

Getting Module not found: Error: Can't resolve 'fs' error with plugin-cloud-storage and s3-adapter

```
ERROR in ./node_modules/.pnpm/@payloadcms+plugin-cloud-storage@1.1.1_@aws-sdk+client-s3@3.454.0_@aws-sdk+lib-storage@3.454.0_payload@2.1.1/node_modules/@payloadcms/plugin-cloud-storage/dist/adapters/s3/handleUpload.js 140:27-40
Module not found: Error: Can't resolve 'fs' in '...\node_modules\.pnpm\@payloadcms+plugin-cloud-storage@1.1.1_@aws-sdk+client-s3@3.454.0_@aws-sdk+lib-storage@3.454.0_payload@2.1.1\node_modules\@payloadcms\plugin-cloud-storage\dist\adapters\s3'
```

### To Reproduce

1- npx create-payload-app@latest reproduce-error-s3 -t blank --use-pnpm
using postgress
2- pnpm add @payloadcms/plugin-cloud-storage
3- pnpm add @aws-sdk/client-s3 @aws-sdk/lib-storage aws-crt
4- optional. add s3 related variables to .env. Not necessary to reproduce the error.
5- adding build config for plugin-cloud-storage & s3-adapter by instructions. (error happens after step 5)
https://github.com/payloadcms/payload/tree/c10db332cdeb3f702d378b0649be20eb789c13e1/packages/plugin-cloud-storage#s3-adapter

```javascript
//payload.config.ts
...
import { cloudStorage } from '@payloadcms/plugin-cloud-storage'
import { s3Adapter } from '@payloadcms/plugin-cloud-storage/s3'

const adapter = s3Adapter({
  config: {
    credentials: {
      accessKeyId: process.env.S3_ACCESS_KEY_ID,
      secretAccessKey: process.env.S3_SECRET_ACCESS_KEY,
    },
    region: process.env.S3_REGION,
  },
  bucket: process.env.S3_BUCKET,
})
...

plugins: [
...
    cloudStorage({
      collections: {
        "my-collection-slug": {
          adapter: adapter,
        },
      },
    }),
  ],
...
```

### Payload Version

2.1.1

### Adapters and Plugins

db-postgres, plugin-cloud-storage, S3 Adapter

**Comment by denolfe** [comment]
@haliacmon Can you give the beta package a try? `pnpm add @payloadcms/plugin-cloud-storage@beta`

**Comment by ericuldall** [comment]
+1 Webpack 5 also not playing nice on my builds.
I've tried numerous fixes, polyfill libraries, installing fallbacks... 
Nothing seems to work and I would love to continue using the cloud storage plugin.

**Comment by ericuldall** [comment]
Wondering if we can just back port webpack to v4??

**Comment by ericuldall** [comment]
Another note, I tried moving to viteBundler, that was full of errors, though I'd love to move over to that once I have some more time to troubleshoot that.

**Comment by ericuldall** [comment]
@denolfe does your fix only address issues with pnpm? I just use npm.

**Comment by denolfe** [comment]
@haliacmon Just published a new beta to try. 

@ericuldall We've identified some issues w/ the vite bundler with this package that are yet to be resolved 👍

**Comment by denolfe** [comment]
> @denolfe does your fix only address issues with pnpm? I just use npm.

I would not expect any issues w/ npm.

**Comment by ericuldall** [comment]
@denolfe well that's good to know, but I can't even get the vite bundler to work on a blank install :-/

**Comment by denolfe** [comment]
@ericuldall Sounds like it warrants a separate issue.

**Comment by ericuldall** [comment]
> @ericuldall Sounds like it warrants a separate issue.

Yes, I'll tackle that in the main repo when the time comes. For now i'm trying to move fast and just get my webpack build working with gcs again :D

**Comment by haliacmon** [comment]
@denolfe Now I get single "Can't resolve 'fs' " error. 

I have tried same steps with npm but the result was same.
If there is anything I can do to help diagnose, test etc. let me know.

```shell
webpack built 4ac35407a2a750e5e806 in 4227ms
ERROR in ./node_modules/.pnpm/@payloadcms+plugin-cloud-storage@1.1.2-beta.1_@aws-sdk+client-s3@3.454.0_@aws-sdk+lib-storage@3.454.0_payload@2.1.1/node_modules/@payloadcms/plugin-cloud-storage/dist/adapters/s3/handleUpload.js 136:49-62
Module not found: Error: Can't resolve 'fs' in 'D:\_code2\Azak\payload-reproduction\node_modules\.pnpm\@payloadcms+plugin-cloud-storage@1.1.2-beta.1_@aws-sdk+client-s3@3.454.0_@aws-sdk+lib-storage@3.454.0_payload@2.1.1\node_modules\@payloadcms\plugin-cloud-storage\dist\adapters\s3'

webpack compiled with 1 error
```

**Comment by denolfe** [comment]
This is quite confusing as we tested your exact repo to make the fix.

**Comment by haliacmon** [comment]
I'm new to payload cms, just started to discover it. I might misunderstand some terminology. I dont have a slug yet... I think.
I just did "npx create-payload-app@latest reproduce-error-s3 -t blank --use-pnpm" and set it up with postgress. then opened localhost:3000 and created admin account. So in the collections I only have "Users" and one user, me/admin. So far zero errors.
![image](https://github.com/payloadcms/payload/assets/7809300/5edbe7d3-9850-4afe-809e-253982a116a0)

but when I tried to confiqure cloud-storage-plugin & S3, I got Can't resolve 'fs' error.

**Comment by haliacmon** [comment]
I found a a workaround.

    webpack: (config) => {
      config.resolve.fallback.fs = false;

      return config;
    },
If I add this to buildConfig() error goes away

```javascript
export default buildConfig({
  admin: {
    user: Users.slug,
    bundler: webpackBundler(),
    webpack: (config) => {
      config.resolve.fallback.fs = false;

      // if (config.resolve?.fallback)
      // config.resolve.fallback = { ...config.resolve.fallback, fs: false };

      return config;
    },
    // bundler: viteBundler(),
  },
```

**Comment by ericuldall** [comment]
> I found a a workaround.
> 
> ```
> webpack: (config) => {
>   config.resolve.fallback.fs = false;
> 
>   return config;
> },
> ```
> 
> If I add this to buildConfig() error goes away
> 
> ```js
> export default buildConfig({
>   admin: {
>     user: Users.slug,
>     bundler: webpackBundler(),
>     webpack: (config) => {
>       config.resolve.fallback.fs = false;
>       
>       // This is to get rid of typescript error
>       // if (config.resolve?.fallback)
>       // config.resolve.fallback = { ...config.resolve.fallback, fs: false };
> 
>       return config;
>     },
>   },
> ...
> ```

Your mileage may vary by disabling those features. I've been trying similar stuff. something else to note if you want to preserve + override the webpack config you can do it like this:
```
  webpack: (config) => ({
    ...config,
    resolve: {
      ...config.resolve,
      fallback: {
        ...config.resolve.fallback,
        fs: false
      }
    }    
  }),
```

**Comment by Lismont** [comment]
> I found a a workaround.
> 
> ```
> webpack: (config) => {
>   config.resolve.fallback.fs = false;
> 
>   return config;
> },
> ```
> 
> If I add this to buildConfig() error goes away
> 
> ```js
> export default buildConfig({
>   admin: {
>     user: Users.slug,
>     bundler: webpackBundler(),
>     webpack: (config) => {
>       config.resolve.fallback.fs = false;
>       
>       // This is to get rid of typescript error
>       // if (config.resolve?.fallback)
>       // config.resolve.fallback = { ...config.resolve.fallback, fs: false };
> 
>       return config;
>     },
>   },
> ...
> ```

thanks this helped me (I bumped into the same problem)

**Comment by tyteen4a03** [comment]
Can confirm I'm also running into this issue.

**Comment by d717an** [comment]
Using `npx create-payload-app@latest` with the website demo on MacOS, I was also running into these webpack errors for `fs`, `os`, and `util` (a total of 12 errors, most for `fs`).

Switching to `@payloadcms/plugin-cloud-storage@beta` brought this down to a single webpack error: `Can't resolve 'fs' in '~/payload-s3/node_modules/@payloadcms/plugin-cloud-storage/dist/adapters/s3`.

Adding the fallback object to `payload.config.ts` allowed me to finally get back into the `/admin` panel.

```
  fallback: {
    ...config.resolve.fallback,
    fs: false,
  },
```

**Comment by BarrySaikSoundry** [comment]
For anyone landing here - 
Don't forget to do this:
https://payloadcms.com/docs/upload/overview

**Comment by BarrySaikSoundry** [comment]
For anyone landing here - 
Don't forget to do this:
https://payloadcms.com/docs/upload/overview

**Comment by huseyinonalcom** [comment]
I've come across this issue as well, I've used the node-polyfill-webpack-plugin package to solve it.

payload.config.ts:

`
...
import { cloudStorage } from "@payloadcms/plugin-cloud-storage";
import { s3Adapter } from "@payloadcms/plugin-cloud-storage/s3";
import NodePolyfillPlugin from "node-polyfill-webpack-plugin";

const s3ad = s3Adapter({
  config: {
    credentials: {
      accessKeyId: process.env.S3_ACCESS_KEY_ID,
      secretAccessKey: process.env.S3_SECRET_ACCESS_KEY,
    },
    region: process.env.S3_REGION,
  },
  bucket: process.env.S3_BUCKET,
});

export default buildConfig({
  admin: {
    user: Users.slug,
    bundler: webpackBundler(),
    webpack: (config) => ({
      ...config,
      plugins: [...config.plugins, new NodePolyfillPlugin()],
    }),
  },
...
  plugins: [
    cloudStorage({
      collections: {
        logos: {
          adapter: s3ad,
        },
        productImages: {
          adapter: s3ad,
        },
      },
    }),
  ],
 ...
`

here's are my dependencies:

`{
...
  "dependencies": {
    "@aws-sdk/client-s3": "^3.582.0",
    "@aws-sdk/lib-storage": "^3.582.0",
    "@payloadcms/bundler-webpack": "^1.0.0",
    "@payloadcms/db-postgres": "^0.x",
    "@payloadcms/plugin-cloud": "^3.0.0",
    "@payloadcms/plugin-cloud-storage": "^1.1.2",
    "@payloadcms/richtext-slate": "^1.0.0",
    "aws-crt": "^1.21.2",
    "cross-env": "^7.0.3",
    "dotenv": "^8.2.0",
    "express": "^4.19.2",
    "node-polyfill-webpack-plugin": "^3.0.0",
    "payload": "^2.0.0"
  },
  "devDependencies": {
    "@types/express": "^4.17.9",
    "copyfiles": "^2.4.1",
    "nodemon": "^2.0.6",
    "ts-node": "^9.1.1",
    "typescript": "^4.8.4"
  }
}
`

here's the upload config of a collection:

`
...
const Logos: CollectionConfig = {
  slug: "logos",
  upload: {
    disableLocalStorage: true,
    staticURL: "https://hocv1testingcdn.cloudfront.net",
    staticDir: "logos",
    mimeTypes: ["image/*"],
  },
...
`

I haven't done any extensive testing, but the admin panels seems to work perfectly.

**Comment by AlessioGr** [comment]
> Replacing the placeholder 'my-collection-slug' with something that exists like `media` completely fixed the problem.

@d717an is correct. This issue, as reproduced here, only occurs if there is an invalid plugin-cloud-storage plugin configuration, e.g. an invalid collection slug.

This has now been fixed by https://github.com/payloadcms/payload/pull/6786 which will throw a proper error if the plugin config is invalid! This will now work as expected, as payload is not supposed to start up with this plugin enabled, while its config is invalid. More info in that PR description
